### PR TITLE
TS-4653: esi plugin - disable HTTP_COOKIE variable by default and imp…

### PR DESCRIPTION
**Pull request by shukitchan**
…lement a whitelist mechanism to allow the specified cookies for it

Original code and idea contributed by Chris Rohlf (chris.rohlf@gmail.com)

@bryancall / @jpeach , pls help to review. Thanks a lot.

### CODE REVIEW

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/322/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/428/ for details.

**Comment by shukitchan** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/429/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/430/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/323/ for details.

**Comment by struct** [comment]
Looks like this version of the patch leaves out wildcard support https://github.com/apache/trafficserver/pull/798/files#diff-a6bc69fbcc52f7fff04507b2a650ac4fR370 I think its sane to support that functionality as long as its not the default

**Comment by shukitchan** [comment]
@struct , i can add the wildcard support tonight.

e.g. whitelistCookie *

That will mean we allow all cookies when we are using HTTP_COOKIE esi variables

**Comment by shukitchan** [comment]
@zwoop can you take a look at this as well?

**Comment by shukitchan** [comment]
wildcard support is added.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/431/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/324/ for details.

**Comment by shukitchan** [comment]
@bryancall / @jpeach / @zwoop any comments here? thanks.

File: plugins/experimental/esi/lib/Variables.cc [hunk 1]
```diff
@@ -357,9 +357,26 @@ Variables::_parseCookieString(const char *str, int str_len)
...
+    std::string v = iter->name;
+    size_t eq     = v.find("=");
+
+    if (eq != std::string::npos) {
+      v = v.substr(0, eq);
+    }
+
+    bool found = false;
+    for (Utils::HeaderValueList::iterator lz = _whitelistCookies.begin(); lz != _whitelistCookies.end(); ++lz) {
+      std::string c = *lz;
```
Comment (zwoop) [inline plugins/experimental/esi/lib/Variables.cc]: Maybe this is the norm for the esi, but why create this intermediary std::string c? That seems inefficient and unnecessary ? Can't you just use the *lz ?

File: plugins/experimental/esi/esi.cc [hunk 2]
```diff
@@ -61,6 +61,7 @@ struct OptionInfo {
 };
 
 static HandlerManager *gHandlerManager = NULL;
+static Utils::HeaderValueList gWhitelistCookies;
```
Comment (zwoop) [inline plugins/experimental/esi/esi.cc]: I'm not familiar with the ESI plugin, but there's only one "config" ever? It sort of feels that the whitelist ought to be associated with the handler (config) no? Even if not supported now, maybe later you would want to?

Comment (shukitchan) [inline plugins/experimental/esi/esi.cc]: yeah... i can clean it up later on

**Comment by shukitchan** [comment]
@zwoop , i just did the cleanup. Good to merge?

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/442/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/336/ for details.

**Comment by shukitchan** [comment]
@sudheerv can you help to do a review for me?

File: plugins/experimental/esi/lib/Variables.cc [hunk 3]
```diff
@@ -357,9 +357,25 @@ Variables::_parseCookieString(const char *str, int str_len)
   AttributeList cookies;
   Utils::parseAttributes(str, str_len, cookies, ";,");
   for (AttributeList::iterator iter = cookies.begin(); iter != cookies.end(); ++iter) {
-    _insert(_dict_data[HTTP_COOKIE], string(iter->name, iter->name_len), string(iter->value, iter->value_len));
-    _debugLog(_debug_tag, "[%s] Inserted cookie with name [%.*s] and value [%.*s]", __FUNCTION__, iter->name_len, iter->name,
-              iter->value_len, iter->value);
+    std::string v = iter->name;
```
Comment (bryancall) [inline plugins/experimental/esi/lib/Variables.cc]: The variable names are not chosen well.  Can you please not use 1 or 2 characters for variable names.

**Comment by bryancall** [comment]
@shukitchan Overall it looks good.  I had one minor comment.

**Comment by shukitchan** [comment]
updated just now. will merge after squashing commits into one. thanks.

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/353/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/457/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/352/ for details.

**Pull request merged**